src/thrift-client/TgtInterfaceImpl.cpp

```diff
@@ -848,6 +844,87 @@ void StordVmdk::UpdateStats(Request* reqp) {
 		} else {
 			stats_.truncate_latency_ += latency;
 		}
+	case Request::Type::kSync:
+		/* No stats requires for sync as it will not fails in thrift */
+		break;
+	}
+}
+
+std::ostream& operator << (std::ostream& os, const RequestBase::Type type) {
+	switch (type) {
+	case RequestBase::Type::kRead:
+		os << "read";
+		break;
+	case RequestBase::Type::kWrite:
+		os << "write";
+		break;
+	case RequestBase::Type::kWriteSame:
+		os << "writesame";
+		break;
+	case RequestBase::Type::kTruncate:
+		os << "truncate";
+		break;
+	case RequestBase::Type::kSync:
+		os << "sync";
+		break;
+	}
+	return os;
+}
+
+std::ostream& operator << (std::ostream& os, const SyncRequest& request) {
+	os << "ID " << request.id
+		<< " type " << request.type
+		<< " privatep " << request.privatep
+		<< " offset " << request.offset
+		<< " length " << request.length;
+	return os;
+}
+
+std::ostream& operator << (std::ostream& os, const Request& request) {
+	os << "ID " << request.id
+		<< " type " << request.type
+		<< " privatep " << request.privatep
+		<< " bufferp " << request.bufferp
+		<< " buf_sz " << request.buf_sz
+		<< " offset " << request.offset
+		<< " length " << request.length;
+	return os;
+}
+
+RequestBase::RequestBase(RequestID id, Type t, const void* privatep,
+	uint64_t length, int64_t offset) : id(id), type(t), privatep(privatep),
+	length(length), offset(offset) {
+}
+
+RequestBase::~RequestBase() {
+}
+
+const RequestBase::Type& RequestBase::GetType() const noexcept {
+		return type;
+}
+
+bool RequestBase::IsOverlapped(uint64_t req_offset,
+	uint64_t req_length) const noexcept {
+	return ((length - req_offset > 0) && (req_length - offset > 0));
+}
+
+Request::Request(RequestID id, Type t, const void* privatep, char *bufferp,
+	int32_t buf_sz, uint64_t length, int64_t offset) : RequestBase(id, t,
+	privatep, length, offset), bufferp(bufferp), buf_sz(buf_sz) {
+}
+
+Request::~Request() {
+}
+
+SyncRequest::SyncRequest(RequestID id, Type t, const void* privatep,
+	uint64_t length, int64_t offset, StordVmdk* vmdkp) :
+	RequestBase(id, t, privatep, length, offset), vmdkp(vmdkp) {
+}
+
+SyncRequest::~SyncRequest() {
+	for (auto write_req : write_pending) {
+		Request *reqp = reinterpret_cast<Request *>(write_req.lock().get());
```

Comment:
weak_ptr.lock() does not guarantee returned values is shared_ptr.

Reply:
Will update to shared ptr. so, locking will not be required.